File: components/si-lang-js/src/remote_function.ts
export interface RemoteFunctionRequestResolver {
  kind: "resolver";
  code: string;
  containerImage: string;
  containerTag: string;
}

export type RemoteFunctionRequest = RemoteFunctionRequestResolver;

export interface RemoteFunctionOutputLine {
  stream: "stdout" | "stderr";
  level: "debug" | "info" | "warn" | "error";
  group: string;
  message: string;
}

export interface RemoteFunctionResultFailure {
  status: "failure";
  kind: string;
  error: {
    message: string;
  };
  data?: never;
}

export interface RemoteFunctionResultResolver {
  status: "success";
  kind: "resolver";
  error?: never;
  data: unknown;
}

export type RemoteFunctionResult =
  | RemoteFunctionRequestResolver
  | RemoteFunctionResultFailure;

File: components/si-model/src/remote_function.rs
use serde::{Deserialize, Serialize};
use thiserror::Error;

#[derive(Error, Debug)]
pub enum RemoteFunctionError {}

pub type RemoteFunctionResult<T> = Result<T, RemoteFunctionError>;

#[derive(Deserialize, Serialize, Debug, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum RemoteFunctionKind {
    Resolver,
}

#[derive(Deserialize, Serialize, Debug, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct RemoteFunctionRequest {
    pub kind: RemoteFunctionKind,
    pub code: String,
    pub container_image: String,
    pub container_tag: String,
}
